store/store_download.go

```diff
@@ -253,6 +259,109 @@ func downloadReqOpts(storeURL *url.URL, cdnHeader string, opts *DownloadOptions)
 	return &reqOptions
 }
 
+type downloadTimeoutError struct {
+	// for testing, to inspect speed
+	Speed float64
+}
+
+func (e *downloadTimeoutError) Error() string {
+	return fmt.Sprintf("download timed out")
+}
+
+// implements io.Writer interface
+// XXX: move to osutil?
+type TransferSpeedMonitoringWriter struct {
+	mu sync.Mutex
+
+	measureTimeWindow   time.Duration
+	minDownloadSpeedBps float64
+
+	ctx context.Context
+
+	// internal state
+	start   time.Time
+	written int
+	cancel  func()
+	err     error
+
+	// for testing
+	measuredWindows int
+}
+
+// NewTransferSpeedMonitoringWriterAndContext returns an io.Writer that measures write speed
+// in measureTimeWindow windows an cancels the operation if minDownloadSeepdBps
+// is not achieved.
+// Monitor() must be called to start actual measurement.
+func NewTransferSpeedMonitoringWriterAndContext(origCtx context.Context, measureTimeWindow time.Duration, minDownloadSpeedBps float64) (*TransferSpeedMonitoringWriter, context.Context) {
+	ctx, cancel := context.WithCancel(origCtx)
+	w := &TransferSpeedMonitoringWriter{
+		measureTimeWindow:   measureTimeWindow,
+		minDownloadSpeedBps: minDownloadSpeedBps,
+		ctx:                 ctx,
+		cancel:              cancel,
+	}
+	return w, ctx
+}
+
+func (w *TransferSpeedMonitoringWriter) reset() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.written = 0
+	w.start = time.Now()
+	w.measuredWindows++
+}
+
+func (w *TransferSpeedMonitoringWriter) checkSpeed(min float64) bool {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	d := time.Now().Sub(w.start)
+	s := float64(w.written) / d.Seconds()
+	ok := s >= min
+	if !ok {
+		w.err = &downloadTimeoutError{Speed: s}
+	}
+	return ok
+}
+
+// Monitor starts a new measurement for write operations and returns a quit
+// channel that should be closed by the caller to finish the measurement.
+func (w *TransferSpeedMonitoringWriter) Monitor() (quit chan bool) {
+	quit = make(chan bool)
+	w.reset()
+	go func() {
+		for {
+			select {
+			case <-time.After(w.measureTimeWindow):
+				if !w.checkSpeed(w.minDownloadSpeedBps) {
+					w.cancel()
+					return
+				}
+				// reset the measurement every downloadSpeedMeasureWindow,
+				// we want average speed per second over short period,
```

Comment:
(nitpick) maybe be more precise here and say something like `we want average speed per second over the mesure time window,`